Replace all usages of `BTreeMap` with `RBTree`.

wedsonaf opened the pull request:
This is based on #400.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

ksquirrel commented:
Review of `f3b6ed1bf410`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`6e35e48`](https://github.com/Rust-for-Linux/linux/commit/6e35e48f2440995acaffc6ec02f988e4ae63be8c): Looks fine!
  - ✔️ Commit [`f3b6ed1`](https://github.com/Rust-for-Linux/linux/commit/f3b6ed1bf410360dda7f765d9af2555168e772cc): Looks fine!

<!-- Review f3b6ed1bf410360dda7f765d9af2555168e772cc -->

ojeda commented on drivers/android/process.rs:
```diff
@@ -260,25 +261,25 @@ impl NodeRefInfo {
-    // for the other fields; we can also get rid of allocations in BTreeMap once we replace it.
+    // TODO: For now this a mutex because we have allocations in RangeAllocator while holding the
+    // lock. We may want to split up the process state at some point to use a spin lock for the
+    // other field
```
Missing period, or there is some more explanation that you did not finish writing?

ojeda replied:
i.e. I ask not because the period is important now, but just in case it was something else.

wedsonaf replied:
There's a missing "s" and period. :) Thanks for catching it!

nbdd0121 commented on drivers/android/process.rs:
```diff
@@ -855,12 +872,11 @@ impl FileOperations for Process {
 
         // Remove all death notifications from the nodes (that belong to a different process).
-        for info in node_refs.values_mut() {
+        for (_, info) in &mut node_refs {
```
Probably better to add `values`, `values_mut` to RBTree instead?

wedsonaf replied:
Probably better, yes. But not a priority at the moment as the code seems readable enough.

nbdd0121 replied:
It should be a straightforward addition if `impl` syntax is used.

```rust
pub fn keys(&self) -> impl Iterator<Item = &'_ K> {
    self.iter().map(|(k, _)| k)
}

pub fn values(&self) -> impl Iterator<Item = &'_ V> {
    self.iter().map(|(_, v)| v)
}

pub fn values_mut(&mut self) -> impl Iterator<Item = &'_ mut V> {
    self.iter_mut().map(|(_, v)| v)
}
```

wedsonaf replied:
Oh, nice. I didn't know about this syntax, I thought I had to define a new type. I was dreading that and writing more documentation.

This should be straightforward enough. I'll update this PR and the rbtree one. Thanks!

nbdd0121 commented on drivers/android/process.rs:
```diff
@@ -872,19 +888,17 @@ impl FileOperations for Process {
-        let mut nodes = BTreeMap::new();
-        swap(&mut inner.threads, &mut threads);
-        swap(&mut inner.nodes, &mut nodes);
+        let threads = replace(&mut inner.threads, RBTree::new());
```
`core::mem::take`?

wedsonaf replied:
I had `swap` elsewhere and @bjorn3 suggested `replace`. Now that I finally replaced all instances of `swap` I learn about `take`.

I'll make this change.

ksquirrel commented:
Review of `4f352b44db23`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`fbe2d34`](https://github.com/Rust-for-Linux/linux/commit/fbe2d3444ab2d510947ffe6f5f86780c7190a61c): Looks fine!
  - ✔️ Commit [`4f352b4`](https://github.com/Rust-for-Linux/linux/commit/4f352b44db2326830da8752007ea4fc5442d7372): Looks fine!

<!-- Review 4f352b44db2326830da8752007ea4fc5442d7372 -->

wedsonaf commented:
### v1 -> v2

* Fixed comment about splitting locks.
* Use `take` instead of `replace`.
* Explicitly implement `Default` for `RBTree` because derivation was requiring key/value types to implement `Default` as well.
* Use `keys`, `values`, and `values_mut`.
* Rebased to the latest #400

ksquirrel commented:
Review of `e3b5b78beb80`:
  - ✔️ Commit [`e3b5b78`](https://github.com/Rust-for-Linux/linux/commit/e3b5b78beb8001e4ae55770a0169b2737bfa7bc4): Looks fine!

<!-- Review e3b5b78beb8001e4ae55770a0169b2737bfa7bc4 -->
